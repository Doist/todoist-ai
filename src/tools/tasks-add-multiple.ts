import type { Task } from "@doist/todoist-api-typescript";
import { z } from "zod";
import type { TodoistTool } from "../todoist-tool.js";
import { mapTask } from "./shared.js";

const TaskSchema = z.object({
	content: z.string().min(1).describe("The content of the task to create."),
	description: z.string().optional().describe("The description of the task."),
	priority: z
		.number()
		.int()
		.min(1)
		.max(4)
		.optional()
		.describe("The priority of the task (1-4)."),
	dueString: z
		.string()
		.optional()
		.describe("The due date for the task, in natural language."),
});

const ArgsSchema = {
	projectId: z
		.string()
		.optional()
		.describe("The project ID to add the tasks to."),
	sectionId: z
		.string()
		.optional()
		.describe("The section ID to add the tasks to."),
	parentId: z
		.string()
		.optional()
		.describe("The parent task ID (for subtasks)."),
	tasks: z.array(TaskSchema).min(1).describe("The array of tasks to add."),
};

const tasksAddMultiple = {
	name: "tasks-add-multiple",
	description: "Add one or more tasks to a project, section, or parent.",
	parameters: ArgsSchema,
	async execute(args, client) {
		const { projectId, sectionId, parentId, tasks } = args;
		const newTasks: Task[] = [];
		for (const task of tasks) {
			const taskArgs = { ...task, projectId, sectionId, parentId };
			newTasks.push(await client.addTask(taskArgs));
		}
		return newTasks.map(mapTask);
	},
} satisfies TodoistTool<typeof ArgsSchema>;

export { tasksAddMultiple };
