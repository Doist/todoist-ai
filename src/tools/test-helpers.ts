import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const
