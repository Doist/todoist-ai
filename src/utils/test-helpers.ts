import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'
import { getToolOutput } from '../mcp-helpers'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
    duration: string | null
    responsibleUid: string | null
    assignedByUid: string | null
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        url: 'https://todoist.com/sections/section-123',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        duration: null,
        responsibleUid: null,
        assignedByUid: null,
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Extracts the text content from a tool output for snapshot testing.
 * This allows tests to match against just the text content while tools return structured output.
 */
export function extractTextContent(toolOutput: unknown): string {
    if (typeof toolOutput === 'string') {
        return toolOutput
    }

    if (typeof toolOutput === 'object' && toolOutput !== null && 'content' in toolOutput) {
        const output = toolOutput as { content: unknown }
        if (
            Array.isArray(output.content) &&
            output.content[0] &&
            typeof output.content[0] === 'object' &&
            output.content[0] !== null &&
            'type' in output.content[0] &&
            'text' in output.content[0] &&
            output.content[0].type === 'text'
        ) {
            return output.content[0].text as string
        }
    }

    throw new Error('Expected tool output to have text content')
}

/**
 * Extracts the structured content from a tool output for testing.
 * This handles both the new `structuredContent` field and legacy JSON-encoded content.
 */
export function extractStructuredContent(
    output: ReturnType<typeof getToolOutput>,
): Record<string, unknown> {
    // Check for new structuredContent field first
    if ('structuredContent' in output && typeof output.structuredContent === 'object') {
        return output.structuredContent as Record<string, unknown>
    }

    // Fall back to checking for JSON content in the content array
    if ('content' in output && Array.isArray(output.content)) {
        for (const item of output.content) {
            if (
                typeof item === 'object' &&
                item !== null &&
                'type' in item &&
                'text' in item &&
                item.type === 'text' &&
                item.mimeType === 'application/json'
            ) {
                return JSON.parse(item.text) as Record<string, unknown>
            }
        }
    }

    throw new Error('Expected tool output to have structured content')
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const

/**
 * Fixed date for consistent test snapshots.
 * Use this instead of new Date() in tests to avoid snapshot drift.
 */
export const TODAY = '2025-08-17' as const
