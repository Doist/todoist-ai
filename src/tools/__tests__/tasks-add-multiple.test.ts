import type { Task, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import { tasksAddMultiple } from '../tasks-add-multiple.js'
import { createMockTask } from '../test-helpers.js'

// Mock the Todoist API
const mockTodoistApi = {
    addTask: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

describe('tasks-add-multiple tool', () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('adding multiple tasks', () => {
        it('should add multiple tasks and return mapped results', async () => {
            // Mock API responses extracted from recordings (Task type)
            const mockApiResponse1: Task = createMockTask({
                id: '8485093748',
                content: 'First task content',
                url: 'https://todoist.com/showTask?id=8485093748',
                addedAt: '2025-08-13T22:09:56.123456Z',
            })

            const mockApiResponse2: Task = createMockTask({
                id: '8485093749',
                content: 'Second task content',
                description: 'Task description',
                labels: ['work', 'urgent'],
                childOrder: 2,
                priority: 2,
                url: 'https://todoist.com/showTask?id=8485093749',
                addedAt: '2025-08-13T22:09:57.123456Z',
                due: {
                    date: '2025-08-15',
                    isRecurring: false,
                    lang: 'en',
                    string: 'Aug 15',
                    timezone: null,
                },
            })

            mockTodoistApi.addTask
                .mockResolvedValueOnce(mockApiResponse1)
                .mockResolvedValueOnce(mockApiResponse2)

            const result = await tasksAddMultiple.execute(
                {
                    projectId: '6cfCcrrCFg2xP94Q',
                    tasks: [
                        { content: 'First task content' },
                        {
                            content: 'Second task content',
                            description: 'Task description',
                            priority: 2,
                            dueString: 'Aug 15',
                        },
                    ],
                },
                mockTodoistApi,
            )

            // Verify API was called correctly for each task
            expect(mockTodoistApi.addTask).toHaveBeenCalledTimes(2)
            expect(mockTodoistApi.addTask).toHaveBeenNthCalledWith(1, {
                content: 'First task content',
                projectId: '6cfCcrrCFg2xP94Q',
                sectionId: undefined,
                parentId: undefined,
            })
            expect(mockTodoistApi.addTask).toHaveBeenNthCalledWith(2, {
                content: 'Second task content',
                description: 'Task description',
                priority: 2,
                dueString: 'Aug 15',
                projectId: '6cfCcrrCFg2xP94Q',
                sectionId: undefined,
                parentId: undefined,
            })

            // Verify result is properly mapped
            expect(result).toEqual([
                expect.objectContaining({
                    id: '8485093748',
                    content: 'First task content',
                    description: '',
                    labels: [],
                }),
                expect.objectContaining({
                    id: '8485093749',
                    content: 'Second task content',
                    description: 'Task description',
                    dueDate: '2025-08-15',
                    priority: 2,
                    labels: ['work', 'urgent'],
                }),
            ])
        })

        it('should handle tasks with section and parent IDs', async () => {
            const mockApiResponse: Task = createMockTask({
                id: '8485093750',
                content: 'Subtask content',
                description: 'Subtask description',
                priority: 3,
                sectionId: 'section-123',
                parentId: 'parent-task-456',
                url: 'https://todoist.com/showTask?id=8485093750',
                addedAt: '2025-08-13T22:09:58.123456Z',
            })

            mockTodoistApi.addTask.mockResolvedValue(mockApiResponse)

            const result = await tasksAddMultiple.execute(
                {
                    projectId: '6cfCcrrCFg2xP94Q',
                    sectionId: 'section-123',
                    parentId: 'parent-task-456',
                    tasks: [
                        {
                            content: 'Subtask content',
                            description: 'Subtask description',
                            priority: 3,
                        },
                    ],
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.addTask).toHaveBeenCalledWith({
                content: 'Subtask content',
                description: 'Subtask description',
                priority: 3,
                projectId: '6cfCcrrCFg2xP94Q',
                sectionId: 'section-123',
                parentId: 'parent-task-456',
            })

            expect(result).toEqual([
                expect.objectContaining({
                    id: '8485093750',
                    content: 'Subtask content',
                    description: 'Subtask description',
                    priority: 3,
                    sectionId: 'section-123',
                    parentId: 'parent-task-456',
                    labels: [],
                }),
            ])
        })
    })

    describe('error handling', () => {
        it('should propagate API errors', async () => {
            const apiError = new Error('API Error: Task content is required')
            mockTodoistApi.addTask.mockRejectedValue(apiError)

            await expect(
                tasksAddMultiple.execute({ tasks: [{ content: '' }] }, mockTodoistApi),
            ).rejects.toThrow(apiError.message)
        })

        it('should handle partial failures when adding multiple tasks', async () => {
            const mockApiResponse: Task = createMockTask({
                id: '8485093751',
                content: 'First task content',
                url: 'https://todoist.com/showTask?id=8485093751',
                addedAt: '2025-08-13T22:09:59.123456Z',
            })

            const apiError = new Error('API Error: Second task failed')
            mockTodoistApi.addTask
                .mockResolvedValueOnce(mockApiResponse)
                .mockRejectedValueOnce(apiError)

            await expect(
                tasksAddMultiple.execute(
                    {
                        tasks: [
                            { content: 'First task content' },
                            { content: 'Second task content' },
                        ],
                    },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Second task failed')

            // Verify first task was attempted
            expect(mockTodoistApi.addTask).toHaveBeenCalledTimes(2)
        })
    })
})
