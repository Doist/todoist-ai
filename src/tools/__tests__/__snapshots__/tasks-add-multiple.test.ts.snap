// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`tasks-add-multiple tool adding multiple tasks should add multiple tasks and return mapped results 1`] = `
"Added 2 tasks to specified project.
Tasks:
    First task content • P1 • id=8485093748
    Second task content • due 2025-08-15 • P2 • id=8485093749.
Next:
- Use overview to see your updated project organization"
`;

exports[`tasks-add-multiple tool adding multiple tasks should add tasks with duration 1`] = `
"Added 2 tasks to specified project.
Tasks:
    Task with 2 hour duration • P1 • id=8485093752
    Task with 45 minute duration • P1 • id=8485093753.
Next:
- Use overview to see your updated project organization"
`;

exports[`tasks-add-multiple tool adding multiple tasks should handle tasks with section and parent IDs 1`] = `
"Added 1 task to specified project.
Tasks:
    Subtask content • P3 • id=8485093750.
Next:
- Use overview to see your updated project organization"
`;

exports[`tasks-add-multiple tool next steps logic should suggest overview tool when no hasToday context 1`] = `
"Added 1 task to specified project.
Tasks:
    Regular task • P1 • id=8485093756.
Next:
- Use overview to see your updated project organization"
`;

exports[`tasks-add-multiple tool next steps logic should suggest tasks-list-by-date for today when hasToday is true 1`] = `
"Added 1 task.
Tasks:
    Task due today • due 2025-08-17 • P1 • id=8485093755.
Next:
- Use overview to see your updated project organization"
`;
